core/trino-main/src/main/java/io/trino/operator/HashGenerator.java

```diff
@@ -15,22 +15,15 @@
 
 import io.trino.spi.Page;
 
-import static com.google.common.base.Preconditions.checkState;
-
 public interface HashGenerator
 {
     long hashPosition(int position, Page page);
 
     default int getPartition(int partitionCount, int position, Page page)
     {
         long rawHash = hashPosition(position, page);
-
-        // clear the sign bit
-        rawHash &= 0x7fff_ffff_ffff_ffffL;
-
-        int partition = (int) (rawHash % partitionCount);
-
-        checkState(partition >= 0 && partition < partitionCount);
-        return partition;
+        // This function reduces the 64 bit rawHash to [0, partitionCount) uniformly. It first reduces the rawHash to 32 bit
+        // integer x then normalize it to x / 2^32 * partitionCount to reduce the range of x from [0, 2^32) to [0, partitionCount)
+        return (int) ((Integer.toUnsignedLong(Long.hashCode(rawHash)) * partitionCount) >> 32);
```

Comment:
We probably don't get big enough `partitionCount` for it to matter, but the `>> 32` should be `>>> 32`

Reply:
I don't think we can ever overflow here. 
Even if `Long.hashCode(rawHash) = -1` (has all bits set to 1), `Integer.toUnsignedLong` will return 2^32 -1 and if have` partitionCount = Integer.MAX_VALUE (2^31-1)` the multiplication is gonna be `(2^32-1)*(2^31-1)` which is `< 2^63 -1`.

If we could overflow and get a negative number it would be bad because >>> would carry the sign bit to the result so it would be negative as well (which is against the contract of this method)

Reply:
I think you're right that we can't overflow here, but isn't `>>` the one which preserves sign bit rather than `>>>` ?

Reply:
indeed it is.
let's change it to `>>>` then.

Reply:
BTW, this is how that formula works: https://github.com/prestodb/presto/pull/11832#discussion_r230095251